pytest/lib/peer.py

```diff
@@ -29,17 +30,28 @@ async def send_raw(self, raw_message):
         self.writer.write(raw_message)
         await self.writer.drain()
 
+    # returns None on timeout
     async def recv(self, expected=None):
         while True:
             response_raw = await self.recv_raw()
+            if response_raw is None:
+                return None
+
             response = BinarySerializer(schema).deserialize(
                 response_raw, PeerMessage)
 
             if expected is None or response.enum == expected:
                 return response
 
     async def recv_raw(self):
-        length = await self.reader.read(4)
+        try:
+            length = await asyncio.wait_for(self.reader.read(4), 5)
+        except concurrent.futures.TimeoutError:
+            return None
```

Comment:
Instead of adding a timeout here I think it will be better to add a timeout in places that calls `recv`. See comment in repro_2916.py for more details.